docs/proposals/volume-idempotency.md

```diff
@@ -0,0 +1,229 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<strong>
+The latest release of this document can be found
+[here](http://releases.k8s.io/release-1.1/docs/proposals/volume-idempotency.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+## Volume plugins and idempotency
+
+Currently, volume plugins have a `SetUp` method which is called repeatedly in the kubelet sync loop.
+Implementers of plugins must implement idempotency checks in `SetUp` to avoid doing duplicate work.
+In order to simplify life for developers of plugins and correctly support idempotent application
+of ownership management to volumes, we should externalize these idempotency checks and perform them
+in the kubelet.
+
+### Types of idempotence checks
+
+There are two types of idempotence checks that volumes perform today:
+
+1.  Checking for the presence of a mount point
+2.  Checking for the presence of a marker file for the volume in the metadata directory for the
+    plugin type
+
+Externalized idempotency management must support both of these types of checks.  The marker file
+method requires that the marker file be created after `SetUp` is run successfully.
+
+### Non-idempotent volumes
+
+The hostPath volume does not implement an idempotency check because its `SetUp` method is a no-op.
+It should be possible to represent that a volume does not have any idempotency concerns.
+
+### Refreshable volume plugins
+
+Some volume plugins are designed to refresh, like the downward API volume plugin.  In order to
+decouple the implementation of refresh and SetUp logic, there should be a Refresh() method that
+plugins can implement and proper machinery to ensure that only plugins that want this need to
+implement it.
```

Comment:
I don't think it's exactly orthogonal, but it's arguable that it's not necessary to accomplish the point of this PR.  I'm willing to leave this out for now as this has gotten quite complex.  Going to scan the rest of your feedback, but I'm planning on removing it now.